pkg/analysis/ssatags/analyzer.go

```diff
@@ -0,0 +1,181 @@
+}
+
+func newAnalyzer(cfg config.SSATagsConfig) *analysis.Analyzer {
+	defaultConfig(&cfg)
+
+	a := &analyzer{
+		cfg: cfg.ListTypeSetUsage,
+	}
+
+	return &analysis.Analyzer{
+		Name:     name,
+		Doc:      "Check that all array types in the API have the SSA tags and the usage of the tags is correct",
+		Run:      a.run,
+		Requires: []*analysis.Analyzer{inspector.Analyzer},
+	}
+}
+
+func (a *analyzer) run(pass *analysis.Pass) (any, error) {
+	inspect, ok := pass.ResultOf[inspector.Analyzer].(inspector.Inspector)
+	if !ok {
+		return nil, kalerrors.ErrCouldNotGetInspector
+	}
+
+	inspect.InspectFields(func(field *ast.Field, stack []ast.Node, jsonTagInfo extractjsontags.FieldTagInfo, markersAccess markers.Markers) {
+		a.checkField(pass, field, markersAccess)
+	})
+
+	return nil, nil //nolint:nilnil
+}
+
+func (a *analyzer) checkField(pass *analysis.Pass, field *ast.Field, markersAccess markers.Markers) {
+	if _, ok := field.Type.(*ast.ArrayType); !ok {
+		return
+	}
+
+	fieldMarkers := markersAccess.FieldMarkers(field)
+	if fieldMarkers == nil {
+		return
+	}
+
+	fieldName := utils.FieldName(field)
+	listTypeMarkers := fieldMarkers.Get(kubebuildermarkers.KubebuilderListTypeMarker)
+
+	if len(listTypeMarkers) == 0 {
+		pass.Report(analysis.Diagnostic{
+			Pos:     field.Pos(),
+			Message: fmt.Sprintf("%s should have a listType marker (atomic, set, or map)", fieldName),
+		})
+
+		return
+	}
+
+	for _, marker := range listTypeMarkers {
+		listType := strings.TrimSpace(marker.Expressions[""])
+		a.checkListTypeMarker(pass, listType, field)
+
+		if listType == listTypeMap {
+			a.checkListTypeMap(pass, fieldMarkers, field)
+		}
+
+		if listType == listTypeSet {
+			a.checkListTypeSet(pass, marker, field)
+		}
+	}
+}
+
+func (a *analyzer) checkListTypeMarker(pass *analysis.Pass, listType string, field *ast.Field) {
+	fieldName := utils.FieldName(field)
+
+	if !validListType(listType) {
+		pass.Report(analysis.Diagnostic{
+			Pos:     field.Pos(),
+			Message: fmt.Sprintf("%s has invalid listType %q, must be one of: atomic, set, map", fieldName, listType),
+		})
+
+		return
+	}
+}
+
+func (a *analyzer) checkListTypeMap(pass *analysis.Pass, fieldMarkers markers.MarkerSet, field *ast.Field) {
+	listMapKeyMarkers := fieldMarkers.Get(kubebuildermarkers.KubebuilderListMapKeyMarker)
+	fieldName := utils.FieldName(field)
+
+	if len(listMapKeyMarkers) == 0 {
+		pass.Report(analysis.Diagnostic{
+			Pos:     field.Pos(),
+			Message: fmt.Sprintf("%s with listType=map must have at least one listMapKey marker", fieldName),
+		})
+	}
+}
+
+func (a *analyzer) checkListTypeSet(pass *analysis.Pass, listTypeMarker markers.Marker, field *ast.Field) {
+	if a.cfg == config.SSATagsListTypeSetUsageIgnore {
+		return
+	}
+
+	fieldName := utils.FieldName(field)
+
+	diagnostic := analysis.Diagnostic{
+		Pos:     field.Pos(),
+		Message: fmt.Sprintf("listType=set is forbidden, use listType=%s or listType=%s instead", listTypeAtomic, listTypeMap),
+	}
```

Comment:
Only when it is a list of objects. A list of string, list of int, should not cause an issue with `listType=set`. 

Personally, I'd like to be able to configure this.

Perhaps like the `nomaps` implementation, we want to have the option of enforcing no sets always, or just for lists of objects?

Reply:
Okay, I understand it. The diagnostic will be reported if the type is a list of objects and the configuration like nomaps is what it should reprot.
It's good way and consistent with nomaps.